File: nafos-core/src/main/java/nafos/bootStrap/handle/http/NsRequest.java
package nafos.bootStrap.handle.http;

import io.netty.buffer.ByteBuf;
import io.netty.handler.codec.http.HttpHeaders;
import io.netty.handler.codec.http.HttpRequest;
import io.netty.handler.codec.http.QueryStringDecoder;
import io.netty.handler.codec.http.cookie.Cookie;
import io.netty.handler.codec.http.cookie.ServerCookieDecoder;
import io.netty.handler.codec.http.multipart.*;
import io.netty.util.CharsetUtil;
import nafos.bootStrap.handle.http.BuildHttpObjectAggregator;
import nafos.core.util.AESUtil;
import nafos.core.util.CastUtil;
import nafos.core.util.JsonUtil;
import nafos.core.util.ObjectUtil;
import org.apache.commons.beanutils.BeanUtils;

import java.io.IOException;
import java.lang.reflect.InvocationTargetException;
import java.net.URLDecoder;
import java.net.URLEncoder;
import java.util.*;

public class NsRequest extends BuildHttpObjectAggregator.AggregatedFullHttpRequest {

    private final String cookieStart = "nafosCookie";

    private String securityCookieId = null;//登陆的cookieId

    private Set<Cookie> cookies = new HashSet<>();//cookieList

    private Map<String, String> requestParams;

    private Map<String, Object> bodyParams;

    NsRequest(HttpRequest request, ByteBuf content, HttpHeaders trailingHeaders) {
        super(request, content, trailingHeaders);
    }


    public String stringQueryParam(String key) {
        return requestParams().get(key);
    }

    public Integer intQueryParam(String key) {
        String value = requestParams().get(key);
        if (value == null) return null;
        return Integer.valueOf(value);
    }

    public Boolean booleanQueryParam(String key) {
        String value = requestParams().get(key);
        if (value == null) return null;
        return Boolean.valueOf(value);
    }

    public Long longQueryParam(String key) {
        String value = requestParams().get(key);
        if (value == null) return null;
        return Long.valueOf(value);
    }


    public String stringBodyParam(String key) {
        Object value = bodyParams().get(key);
        if (value == null) return null;
        return String.valueOf(value);
    }

    public Integer intBodyParam(String key) {
        Object value = bodyParams().get(key);
        if (value == null) return null;
        return Integer.valueOf(value.toString());
    }

    public Boolean booleanBodyParam(String key) {
        Object value = bodyParams().get(key);
        if (value == null) return null;
        return Boolean.valueOf(value.toString());
    }

    public Long longBodyParam(String key) {
        Object value = bodyParams().get(key);
        if (value == null) return null;
        return Long.valueOf(value.toString());
    }


    /**
     * 获取cookie列表
     *
     * @return
     */
    public Set<Cookie> getCookies() {
        if (cookies.isEmpty()) {
            String cookieStr = headers().get("Cookie");
            if (ObjectUtil.isNotNull(cookieStr)) {
                cookies = ServerCookieDecoder.LAX.decode(cookieStr);
            }
        }
        return cookies;
    }

    public String getCookie(String name) {
        Iterator<Cookie> it = getCookies().iterator();
        while (it.hasNext()) {
            Cookie cookie = it.next();
            if (cookie.name().equals(name)) {
                return cookie.value();
            }
        }
        return null;
    }


    public String getNafosCookieId() {
        //H5跨域不能设置cookie问题，暂用此方法解决
        if (ObjectUtil.isNotNull(headers().get(cookieStart)) &&
                !headers().get(cookieStart).equals("undefined") &&
                !headers().get(cookieStart).equals("null") &&
                !headers().get(cookieStart).equals("[object Null]")) {
            try {
                securityCookieId = AESUtil.decrypt(URLDecoder.decode(headers().get(cookieStart)));
                return securityCookieId;
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
        String nafosCookieAes = getCookie(cookieStart);
        if(nafosCookieAes == null){
            return null;
        }
        //其他模拟器，安卓正常走流程
        securityCookieId = AESUtil.decrypt(URLDecoder.decode(nafosCookieAes));
        return securityCookieId;
    }


    /**
     * restful风格的postJSON解析
     */
    private Map<String, Object> bodyParams() {
        if (bodyParams == null) {
            // 处理POST请求
            String strContentType = headers().get("Content-Type");
            strContentType = ObjectUtil.isNotNull(strContentType) ? strContentType.trim() : "";
            if (strContentType.contains("application/json")) {
                bodyParams = getJSONParams();
            } else {
                bodyParams = getFormParams();
            }
        }
        return bodyParams;
    }


    /**
     * 解析from表单数据（Content-Type = x-www-form-urlencoded）,默认格式
     */
    public Map<String, Object> getFormParams() {
        Map<String, Object> params = new HashMap<String, Object>();
        HttpPostRequestDecoder decoder = new HttpPostRequestDecoder(new DefaultHttpDataFactory(false), this);
        List<InterfaceHttpData> postData = decoder.getBodyHttpDatas();
        for (InterfaceHttpData data : postData) {
            if (data.getHttpDataType() == InterfaceHttpData.HttpDataType.Attribute) {
                MemoryAttribute attribute = (MemoryAttribute) data;
                params.put(attribute.getName(), attribute.getValue());
            }
        }
        return params;
    }

    /**
     * 解析json数据（Content-Type = application/json）
     */
    public Map<String, Object> getJSONParams() {
        ByteBuf jsonBuf = content();
        String jsonStr = jsonBuf.toString(CharsetUtil.UTF_8);
        if(jsonStr == null || jsonStr.trim().length() == 0) return new HashMap<>();
        return JsonUtil.jsonToMap(jsonStr);
    }


    /**
     * 解析queryparams
     *
     * @return
     */
    private Map<String, String> requestParams() {
        if (requestParams == null) {
            Map<String, String> map = new HashMap<>();
            QueryStringDecoder decoder = new QueryStringDecoder(uri());
            for (Map.Entry<String, List<String>> entry : decoder.parameters().entrySet()) {
                map.put(entry.getKey(), entry.getValue().get(0));
            }
            requestParams = map;
        }
        return requestParams;
    }

    /**
     * xml风格的postJSON解析
     *
     * @param clazz
     * @return
     */
    private Object xmlJsonEncode(Class<?> clazz) {
        Object fieldObj = null;
        Map<String, String> postMap = new HashMap<>();
        HttpPostRequestDecoder httpPostRequestDecoder = new HttpPostRequestDecoder(this);
        httpPostRequestDecoder.offer(this);
        List<InterfaceHttpData> parmList = httpPostRequestDecoder.getBodyHttpDatas();
        for (InterfaceHttpData parm : parmList) {
            Attribute data = (Attribute) parm;
            try {
                postMap.put(data.getName(), data.getValue());
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        if (!Map.class.isAssignableFrom(clazz)) {
            try {
                BeanUtils.populate(fieldObj, postMap);
            } catch (IllegalAccessException e) {
                e.printStackTrace();
            } catch (InvocationTargetException e) {
                e.printStackTrace();
            }
        } else {
            fieldObj = postMap;
        }
        return fieldObj;
    }

    public Map<String, String> getRequestParams() {
        return requestParams();
    }

    public Map<String, Object> getBodyParams() {
        return bodyParams();
    }

}

File: server/src/main/kotlin/nafos/server/handle/http/NsRequest.kt
package nafos.server.handle.http

import io.netty.buffer.ByteBuf
import io.netty.handler.codec.http.HttpHeaders
import io.netty.handler.codec.http.HttpMethod
import io.netty.handler.codec.http.HttpRequest
import io.netty.handler.codec.http.QueryStringDecoder
import io.netty.handler.codec.http.cookie.Cookie
import io.netty.handler.codec.http.cookie.ServerCookieDecoder
import io.netty.handler.codec.http.multipart.*
import io.netty.util.CharsetUtil
import nafos.server.util.JsonUtil
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.util.HashMap

/**
 * @Description nafos-request
 * @Author      xinyu.huang
 * @Time        2019/11/24 20:05
 */
class NsRequest(
        request: HttpRequest,
        content: ByteBuf,
        trailingHeaders: HttpHeaders?
) : BuildHttpObjectAggregator.AggregatedFullHttpRequest(request, content, trailingHeaders) {

    var cookies = setOf<Cookie>()

    var requestParams: Map<String, String> = mutableMapOf()

    var bodyParams: Map<String, Any> = mutableMapOf()

    var ip: String? = null

    private var httpPostRequestDecoder: HttpPostRequestDecoder? = null

    var filesMap: MutableMap<String, File>? = null
        get() {
            /**
             * @Desc     获取POST body中的文件列表
             * @Author   hxy
             * @Time     2019/12/3 11:43
             */
            synchronized(this) {
                field ?: run {
                    val httpDecoder = HttpPostRequestDecoder(DefaultHttpDataFactory(true), this)
                    httpDecoder.setDiscardThreshold(0)
                    field = mutableMapOf()
                    httpDecoder.bodyHttpDatas.forEach {
                        val fileUpload = (it as FileUpload)
                        val fileName = fileUpload.getFilename()
                        val file = fileUpload.getFile()
                        if (fileName != null && file != null) {
                            field!![fileName] = file
                        }
                    }
                }
            }
            return field
        }


    init {
        //cookies
        val cookieStr = headers().get("Cookie")
        cookieStr?.run {
            cookies = ServerCookieDecoder.LAX.decode(cookieStr)
        }
        //注册requestParams
        val map = mutableMapOf<String, String>()
        val decoder = QueryStringDecoder(uri())
        for (entry in decoder.parameters().entries) {
            map[entry.key] = entry.value[0]
        }
        requestParams = map

        //注册bodyParams
        if(request.method() == HttpMethod.POST){
            var strContentType = headers().get("Content-Type")
            strContentType = strContentType?.trim() ?: ""
            bodyParams = if (strContentType.contains("application/json")) {
                getJsonParams()
            } else {
                getFormParams()
            }
        }
    }


    /***
     *@Description 获取单个cookie
     *@Author      xinyu.huang
     *@Time        2019/11/24 20:16
     */
    fun getCookie(name: String): String? {
        val it = cookies.iterator()
        while (it.hasNext()) {
            val cookie = it.next()
            if (cookie.name() == name) {
                return cookie.value()
            }
        }
        return null
    }

    /**
     * @Desc     转存body文件到某个目录下
     * @Author   hxy
     * @Time     2019/12/3 13:44
     */
    fun transferFileFrom(dirPath: String) {
        filesMap?.forEach { t, u ->
            val file = File(dirPath + t)
            FileInputStream(u).channel.use { inputChannel ->
                FileOutputStream(file).channel.use { outputChannel ->
                    outputChannel.transferFrom(inputChannel, 0, inputChannel.size())
                }
            }
        }
    }


    /***
     *@Description 解析from表单数据（Content-Type = x-www-form-urlencoded）,默认格式
     *@Author      xinyu.huang
     *@Time        2019/11/24 20:18
     */
    internal fun getFormParams(): Map<String, Any> {
        val params = HashMap<String, Any>()
        httpPostRequestDecoder = HttpPostRequestDecoder(DefaultHttpDataFactory(false), this)
        val postData = httpPostRequestDecoder!!.bodyHttpDatas
        for (data in postData) {
            if (data.httpDataType == InterfaceHttpData.HttpDataType.Attribute) {
                val attribute = data as MemoryAttribute
                params[attribute.name] = attribute.value
            }
        }
        return params
    }

    /***
     *@Description 解析json数据（Content-Type = application/json）
     *@Author      xinyu.huang
     *@Time        2019/11/24 21:18
     */
    private fun getJsonParams(): Map<String, Any> {
        val jsonBuf = content()
        val jsonStr = jsonBuf.toString(CharsetUtil.UTF_8)
        return if (jsonStr.isNullOrBlank()) {
            HashMap()
        } else JsonUtil.jsonToMap(jsonStr)
    }


    override fun release(): Boolean {
        if (httpPostRequestDecoder != null) {
            httpPostRequestDecoder!!.destroy()
            httpPostRequestDecoder = null
        }
        return super.release()
    }
}